File: editor.py
class Editor:
    def __init__(self):
        self.buffer:list[bytearray] = [bytearray()]
        self.dirty = False
    
    def padLine(self, target:int):
        self.dirty = True
        target += 1
        self.dirty = True
        if target < len(self.buffer):
            return
        add = target - len(self.buffer)
        self.buffer.extend(bytearray() for _ in range(add))
        return add
    
    def getLine(self, line:int):
        try:
            return self.buffer[line].decode()
        except IndexError:
            return ""
    
    # end is inclusive unlike regular python slice
    def getLines(self, begin=0, end=-1):
        if len(self.buffer) == 0:
            return [""]
        
        if end == -1:
            end = len(self.buffer)-1
        
        selected = [item.decode() for item in self.buffer[begin:end+1]]

        return selected

    def insertLine(self, i:int, text:str):
        self.padLine(i)
        self.buffer.insert(i, bytearray(text,encoding="utf-8"))
    
    def removeLine(self, i:int):
        self.dirty = True
        try:
            self.buffer.pop(i)
        except IndexError:
            pass

    def replaceLine(self, i:int, text:str):
        self.padLine(i)
        self.buffer[i] = bytearray(text,encoding="utf-8")
    
    def load(self, data:bytes):
        self.buffer = [bytearray(line) for line in data.split(b"\n")]

File: src/editor.rs
pub struct Editor {
    buffer:Vec<String>,
    dirty:bool
}

impl Editor {
    fn pad_line(&mut self, target:usize) {
        if target < self.buffer.len() {return};
        self.dirty = true;
        while self.buffer.len() < target {
            self.buffer.push(String::new());
        }
    }

    pub fn insert_line(&mut self, line:usize, text:String) {
        self.pad_line(line);
        self.dirty = true;
        self.buffer.insert(line,text);
    }

    pub fn replace_line(&mut self, line:usize, text:String) {
        self.pad_line(line);
        self.dirty = true;
        if line < self.buffer.len() {
            self.buffer[line] = text;
        } else {
            self.buffer.push(text);
        }
    }

    pub fn delete_line(&mut self, line:usize) {
        self.dirty = true;
        self.buffer.remove(line);
    }

    pub fn get_line(&self, line:usize) -> &str {
        self.buffer[line].as_str()
    }

    pub fn get_lines(&self, begin: usize, end: usize) -> Vec<&str> {
        self.buffer[begin..=end]
            .iter()
            .map(String::as_str)
            .collect()
    }

    pub fn new() -> Editor {
        Editor {
            buffer : vec![String::new()],
            dirty : false
        }
    }

}